lsm303agr/lsm303agr.go

```diff
@@ -0,0 +1,196 @@
+/*
+Package lsm303agr implements a driver for the LSM303AGR,
+ a 3 axis accelerometer/magnetic sensor which is included on BBC micro:bits v1.5.
+
+Datasheet: https://www.st.com/resource/en/datasheet/lsm303agr.pdf
+*/
+
+package lsm303agr // import "tinygo.org/x/drivers/lsm303agr"
+
+import (
+	"machine"
+	"math"
+)
+
+/* for the LSM303AGR object */
+type Device struct {
+	bus            machine.I2C
+	AccelAddress   uint8
+	MagAddress     uint8
+	AccelPowerMode uint8
+	AccelRange     uint8
+	AccelDataRate  uint8
+	MagPowerMode   uint8
+	MagSystemMode  uint8
+	MagDataRate    uint8
+}
+
+/* for configuring LSM303AGR */
+type Configuration struct {
+	AccelPowerMode uint8
+	AccelRange     uint8
+	AccelDataRate  uint8
+	MagPowerMode   uint8
+	MagSystemMode  uint8
+	MagDataRate    uint8
+}
+
+/* create a new LSM303AGR object */
+func New(bus machine.I2C) Device {
+	return Device{bus: bus, AccelAddress: ACCEL_ADDRESS, MagAddress: MAG_ADDRESS}
+}
+
+/* check if LSM303AGR's both sensors are connected */
+func (d *Device) Connected() bool {
+	data1, data2 := []byte{0}, []byte{0}
+	d.bus.ReadRegister(uint8(d.AccelAddress), ACCEL_WHO_AM_I, data1)
+	d.bus.ReadRegister(uint8(d.MagAddress), MAG_WHO_AM_I, data2)
+	return data1[0] == 0x33 && data2[0] == 0x40
+}
+
+/* configure and initialize LSM303AGR */
+func (d *Device) Configure(cfg Configuration) {
+
+	if cfg.AccelDataRate != 0 {
+		d.AccelDataRate = cfg.AccelDataRate
+	} else {
+		d.AccelDataRate = ACCEL_DATARATE_100HZ
+	}
+
+	if cfg.AccelPowerMode != 0 {
+		d.AccelPowerMode = cfg.AccelPowerMode
+	} else {
+		d.AccelPowerMode = ACCEL_POWER_NORMAL
+	}
+
+	if cfg.AccelRange != 0 {
+		d.AccelRange = cfg.AccelRange
+	} else {
+		d.AccelRange = ACCEL_RANGE_2G
+	}
+
+	if cfg.MagPowerMode != 0 {
+		d.MagPowerMode = cfg.MagPowerMode
+	} else {
+		d.MagPowerMode = MAG_POWER_NORMAL
+	}
+
+	if cfg.MagDataRate != 0 {
+		d.MagDataRate = cfg.MagDataRate
+	} else {
+		d.MagDataRate = MAG_DATARATE_10HZ
+	}
+
+	if cfg.MagSystemMode != 0 {
+		d.MagSystemMode = cfg.MagSystemMode
+	} else {
+		d.MagSystemMode = MAG_SYSTEM_CONTINUOUS
+	}
+
+	cmd := []byte{0}
+
+	cmd[0] = byte(d.AccelDataRate<<4 | d.AccelPowerMode | 0x07)
+	d.bus.WriteRegister(uint8(d.AccelAddress), ACCEL_CTRL_REG1_A, cmd)
+
+	cmd[0] = byte(0x80 | d.AccelRange<<4)
+	d.bus.WriteRegister(uint8(d.AccelAddress), ACCEL_CTRL_REG4_A, cmd)
+
+	cmd[0] = byte(0xC0)
+	d.bus.WriteRegister(uint8(d.AccelAddress), TEMP_CFG_REG_A, cmd)
+
+	cmd[0] = byte(0x80 | d.MagPowerMode<<4 | d.MagDataRate<<2 | d.MagSystemMode)
+	d.bus.WriteRegister(uint8(d.MagAddress), MAG_MR_REG_M, cmd)
+
+}
+
+/* read raw acceleration data from all axis, 1000+ = 1g */
+func (d *Device) ReadAcceleration() (x int16, y int16, z int16) {
```

Comment:
I think it would be a lot better to support the same API as the other devices in this category. For example, this function would be better defined as `func (d *Device) ReadAcceleration() (x int32, y int32, z int32)` and return values in µg (micro-gravity).

Reply:
I think currently it returns mg (if I lay the board flat on the table, the z axis reads a bit more than -1000, which is the same result I saw in MakeCode.) So times 1000 should be good?

Reply:
Please check this with the datasheet. Usually it is a power of two, so perhaps it's 1024. The datasheet will tell you exactly. Also, many of these chips can be configured to different sensitivities (for example, 2g, 4g, 8g, 16g for different applications), although initially using the default would be fine.

Reply:
I think the [datasheet](https://www.st.com/resource/en/datasheet/lsm303agr.pdf) is poorly written; there's a lot of info it doesn't revel in there. 

In my experiments each axis always returns a value between +- 32768. The bigger the g sensitivity, the bigger the value scale would be (the observed Earth gravity reading become smaller in scale). I guess this is how it can represent a larger value in the same 2 bytes.

(And in normal or low power mode, only top 12 bits of the 2 bytes would be used, can't find the source now. I didn't implement the full 16-bit high resolution mode.)

The readings in 2g is exactly the same as what I've seen measured in MakeCode (1000~plus). In [MakeCode](https://makecode.microbit.org/reference/input/acceleration) the unit of raw axis data is labeled as mg or milligram. 1,000 mg = 1g (Earth gravity).